notify_listener.go

```diff
@@ -166,27 +172,67 @@ func (l *NotifyListener) processMessage(msg *pgx.Notification) {
 		l.errCh <- err
 	}
 
-	l.processChangeset(event)
+	l.logger.Debugf("processMessage: changeset %d", event.ID)
+	l.handleChangeset(event)
 }
 
-func (l *NotifyListener) processChangeset(event *store.Event) {
-	// NOTE: If the changeset ID is less than the ID of the last processed changeset, skip.
-	// This likely means that we've already processed the changeset when reading from the
-	// changesets table, and the connection is playing back buffered notifications.
-	// (see: `pgx.Conn.notifications`)
-	if l.lastProcessedChangeset != nil && event.ID <= l.lastProcessedChangeset.ID {
-		l.logger.Infof("Skipping changest already processed changeset %s", event.ID)
+// handleChangesets implements a strict ordered buffer queue from unordered
+// input to force changesets to be processed in order.
+func (l *NotifyListener) handleChangeset(event *store.Event) {
+	// TODO: Is this queue a workaround for a bug? It solves the problem, but
```

Comment:
I need to understand if/how these are out order. What this an issue that you've seen running an actual migration or just in the tests?

Reply:
We noticed this in the tests yesterday.  We weren't logging enough to determine if the changesets where out of order. I don't see the logs in Brad's nebula run after these changes, not sure what happened there. But looking at the output, looks like the changeset ID is not matching on the source and target, despite the changes being the same. 
```
{"level":"info","msg":"checking first changeset to changeset id: 31062449","time":"2021-03-08T10:15:37Z"}
{"level":"info","msg":"processing changeset id: 30944972","time":"2021-03-08T10:15:37Z"}
{"level":"error","msg":"source/target rows differ, source: ID: 30945496, Timestamp: 0001-01-01 00:00:00 +0000 UTC, Action: UPDATE, TableName: star, NewValues: {'starid':'jupyterbyocgpu-zpekrdfm', 'instancetype':'custom-gpu-1', 'diskspace':50, 'state':'RUNNING', 'ipaddress':'35.224.125.208', 'snapshotid':'i$jupyterbyocg
pu-1598550017', 'poolid':'jupyterbyocgpu', 'poolversion':8, 'userid':'10738380037', 'createdat':'2021-03-08T08:22:43.447', 'updatedat':'2021-03-08T08:58:30.923', 'provisionkey':'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla', 'provisionedat':'2021-03-08T08:33:22.811', 'workspacefiles':'workspace-1589827581.tar.gz'
, 'masterfiles':null, 'meter':'GPU', 'zone':'us-central1-a', 'keepalivetoken':'lnkqviidy8gb7hfr4nkh4zmsdt5shwqr', 'lock':'n1kluhyg7tu8cl9la9g6br4lbsnxex', 'meta':'{\\'check\\':\\'enrollment\\',\\'poolId\\':\\'jupyterbyocgpu\\',\\'poolIdToUse\\':\\'jupyterbyocgpu\\',\\'courseId\\':null,\\'nanodegreeId\\':956018,\\'atomI
d\\':1003776,\\'labId\\':null,\\'classKey\\':\\'nd320\\',\\'workspaceId\\':\\'r956022c971682xJUPYTERpolbwxla\\',\\'provisionKey\\':\\'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla\\',\\'meterChoice\\':\\'GPU\\',\\'meterToUse\\':\\'GPU\\',\\'coco\\':false,\\'reset\\':false,\\'update\\':false,\\'staging\\':null,\\'d
iskId\\':\\'mldatasets\\',\\'paths\\':[\\'/mnt/disks/data/content//AIHCND/conda-envs/conda_tf2_tf_probability/conda:/opt/conda:ro\\',\\'/mnt/disks/data/content//nih_chest_xray:/data:ro\\'],\\'scriptId\\':null,\\'udacian\\':false}', 'step':null, 'steppercent':0, 'terminating':false}, OldValues: {'starid':'jupyterbyocgpu
-zpekrdfm', 'instancetype':'custom-gpu-1', 'diskspace':50, 'state':'RUNNING', 'ipaddress':'35.224.125.208', 'snapshotid':'i$jupyterbyocgpu-1598550017', 'poolid':'jupyterbyocgpu', 'poolversion':8, 'userid':'10738380037', 'createdat':'2021-03-08T08:22:43.447', 'updatedat':'2021-03-08T08:56:26.064', 'provisionkey':'p:jupy
terbyocgpu$$n:r956022c971682xJUPYTERpolbwxla', 'provisionedat':'2021-03-08T08:33:22.811', 'workspacefiles':'workspace-1589827581.tar.gz', 'masterfiles':null, 'meter':'GPU', 'zone':'us-central1-a', 'keepalivetoken':'lnkqviidy8gb7hfr4nkh4zmsdt5shwqr', 'lock':'n1kluhyg7tu8cl9la9g6br4lbsnxex', 'meta':'{\\'check\\':\\'enrol
lment\\',\\'poolId\\':\\'jupyterbyocgpu\\',\\'poolIdToUse\\':\\'jupyterbyocgpu\\',\\'courseId\\':null,\\'nanodegreeId\\':956018,\\'atomId\\':1003776,\\'labId\\':null,\\'classKey\\':\\'nd320\\',\\'workspaceId\\':\\'r956022c971682xJUPYTERpolbwxla\\',\\'provisionKey\\':\\'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla
\\',\\'meterChoice\\':\\'GPU\\',\\'meterToUse\\':\\'GPU\\',\\'coco\\':false,\\'reset\\':false,\\'update\\':false,\\'staging\\':null,\\'diskId\\':\\'mldatasets\\',\\'paths\\':[\\'/mnt/disks/data/content//AIHCND/conda-envs/conda_tf2_tf_probability/conda:/opt/conda:ro\\',\\'/mnt/disks/data/content//nih_chest_xray:/data:ro
\\'],\\'scriptId\\':null,\\'udacian\\':false}', 'step':null, 'steppercent':0, 'terminating':false} target: ID: 30945519, Timestamp: 0001-01-01 00:00:00 +0000 UTC, Action: UPDATE, TableName: star, NewValues: {'starid':'jupyterbyocgpu-zpekrdfm', 'instancetype':'custom-gpu-1', 'diskspace':50, 'state':'RUNNING', 'ipaddress
':'35.224.125.208', 'snapshotid':'i$jupyterbyocgpu-1598550017', 'poolid':'jupyterbyocgpu', 'poolversion':8, 'userid':'10738380037', 'createdat':'2021-03-08T08:22:43.447', 'updatedat':'2021-03-08T08:58:30.923', 'provisionkey':'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla', 'provisionedat':'2021-03-08T08:33:22.811'
, 'workspacefiles':'workspace-1589827581.tar.gz', 'masterfiles':null, 'meter':'GPU', 'zone':'us-central1-a', 'keepalivetoken':'lnkqviidy8gb7hfr4nkh4zmsdt5shwqr', 'lock':'n1kluhyg7tu8cl9la9g6br4lbsnxex', 'meta':'{\\'check\\':\\'enrollment\\',\\'poolId\\':\\'jupyterbyocgpu\\',\\'poolIdToUse\\':\\'jupyterbyocgpu\\',\\'cou
rseId\\':null,\\'nanodegreeId\\':956018,\\'atomId\\':1003776,\\'labId\\':null,\\'classKey\\':\\'nd320\\',\\'workspaceId\\':\\'r956022c971682xJUPYTERpolbwxla\\',\\'provisionKey\\':\\'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla\\',\\'meterChoice\\':\\'GPU\\',\\'meterToUse\\':\\'GPU\\',\\'coco\\':false,\\'reset\\':
false,\\'update\\':false,\\'staging\\':null,\\'diskId\\':\\'mldatasets\\',\\'paths\\':[\\'/mnt/disks/data/content//AIHCND/conda-envs/conda_tf2_tf_probability/conda:/opt/conda:ro\\',\\'/mnt/disks/data/content//nih_chest_xray:/data:ro\\'],\\'scriptId\\':null,\\'udacian\\':false}', 'step':null, 'steppercent':0, 'terminati
ng':false}, OldValues: {'starid':'jupyterbyocgpu-zpekrdfm', 'instancetype':'custom-gpu-1', 'diskspace':50, 'state':'RUNNING', 'ipaddress':'35.224.125.208', 'snapshotid':'i$jupyterbyocgpu-1598550017', 'poolid':'jupyterbyocgpu', 'poolversion':8, 'userid':'10738380037', 'createdat':'2021-03-08T08:22:43.447', 'updatedat':'
2021-03-08T08:56:26.064', 'provisionkey':'p:jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla', 'provisionedat':'2021-03-08T08:33:22.811', 'workspacefiles':'workspace-1589827581.tar.gz', 'masterfiles':null, 'meter':'GPU', 'zone':'us-central1-a', 'keepalivetoken':'lnkqviidy8gb7hfr4nkh4zmsdt5shwqr', 'lock':'n1kluhyg7tu8cl
9la9g6br4lbsnxex', 'meta':'{\\'check\\':\\'enrollment\\',\\'poolId\\':\\'jupyterbyocgpu\\',\\'poolIdToUse\\':\\'jupyterbyocgpu\\',\\'courseId\\':null,\\'nanodegreeId\\':956018,\\'atomId\\':1003776,\\'labId\\':null,\\'classKey\\':\\'nd320\\',\\'workspaceId\\':\\'r956022c971682xJUPYTERpolbwxla\\',\\'provisionKey\\':\\'p:
jupyterbyocgpu$$n:r956022c971682xJUPYTERpolbwxla\\',\\'meterChoice\\':\\'GPU\\',\\'meterToUse\\':\\'GPU\\',\\'coco\\':false,\\'reset\\':false,\\'update\\':false,\\'staging\\':null,\\'diskId\\':\\'mldatasets\\',\\'paths\\':[\\'/mnt/disks/data/content//AIHCND/conda-envs/conda_tf2_tf_probability/conda:/opt/conda:ro\\',\\'
/mnt/disks/data/content//nih_chest_xray:/data:ro\\'],\\'scriptId\\':null,\\'udacian\\':false}', 'step':null, 'steppercent':0, 'terminating':false}","time":"2021-03-08T10:15:38Z"}
```

Reply:
Run the integration test you see the out-of-order messages occur.

> Storing OUT-OF-ORDER unprocessed record id: 158

Reply:
Why? no idea. But it is happening.

Reply:
Original message by @svysali 

> We noticed this in the tests yesterday.  We weren't logging enough to determine if the changesets where out of order. I don't see the logs in Brad's nebula run after these changes, not sure what happened there. But looking at the output, looks like the changeset ID is not matching on the source and target, despite the changes being the same.